### ISSUE TITLE
Adding "narrow" ResNet models for CIFAR-10

### ISSUE THREAD

**Issue by nshaud** [pytorch/vision#1570]
In the original [Residual Network paper](https://arxiv.org/pdf/1512.03385v1.pdf), the authors use a different layout for ImageNet and CIFAR-10 (see section 4.2). This makes it difficult to reproduce the results since torchvision only include built-in ResNets designed for ImageNet. [Some users](https://discuss.pytorch.org/t/reproducing-resnet-cifar-10-test-error/56558) have been confused by this (see also #1166). There is an [unofficial implementation](https://github.com/akamaster/pytorch_resnet_cifar10) yet users could benefit from having the model available straight from torchvision.

This has been requested (at least in #913 and #1166). I understand that the Pytorch team does not plan to work on this, although since I already have this work done on [my own repo](https://github.com/nshaud/vision/blob/narrow_resnets/torchvision/models/resnet.py), I would gladly open a PR if there is an interest. I am able to reproduce the errors from the paper and which is also compatible with the pretrained models from [akamaster](https://github.com/akamaster/pytorch_resnet_cifar10). This does not change the behaviour of usual ResNets but simply extends the current class to support the CIFAR-10 variant.

**Comment by fmassa** [comment]
Hi,

Thanks for the detailed issue with references, and sorry for the delay in replying.

One thing that we currently don't have in torchvision is pre-trained models for different datasets than ImageNet / COCO / Kinetics.
If we were to provide models for other datasets (like Cifar), we would need to make it clear where the model was trained on. One thing which complicates things even further is that the architectures are also different, so we would need to either create different classes or have additional flags to handle those cases.

I think for the sake of simplicity, we won't be adding models for CIFAR-10. They generally do not transfer well to other tasks, and adds extra complexity on the models (should all models have a CIFAR-10 variant?).

As such, I'm closing this issue, but let me know if you think differently.

**Comment by nshaud** [comment]
Thanks for the feedback. I see where you'r'e coming from but I'd like to argue one last time before I drop the matter :smile:.

Just to be clear, I'm not asking for full support for CIFAR-10 with pretrained models and everything. I am mostly arguing that it would be nice to include both ResNet variants (including the "less useful" one) in Pytorch.

> If we were to provide models for other datasets (like Cifar), we would need to make it clear where the model was trained on. One thing which complicates things even further is that the architectures are also different, so we would need to either create different classes or have additional flags to handle those cases.

My point is that (for ResNet, at least) we don't need extra classes. I agree that it adds a few kwargs to the constructor but it's not worse than currently implemented variants (dilation, wide resnets, various normalization layers, etc.). Even without exposing the model with a `torchvision.models.narrow_resnet` attribute, it would be nice to at least be able to manually define the narrow model using an explicit call to the constructor with the right set of parameters, e.g. `torchvision.models.ResNet([3, 3, 3], option='A')`.

Documenting what dataset has been used for pretraining is indeed a problem although it might not matter: CIFAR-10 is arguably small enough that retraining is acceptable.

> I think for the sake of simplicity, we won't be adding models for CIFAR-10. They generally do not transfer well to other tasks, and adds extra complexity on the models (should all models have a CIFAR-10 variant?).

The scope of this issue is confined to ResNets, I'm not arguing that torchvision should implement CIFAR variants for all models (especially since I am not aware of other models that use different topologies for CIFAR and ImageNet).

Regarding simplicity, I'd say that allowing PyTorch users to replicate the results from the original ResNet paper without too much trouble is a quick (small) win. But if you say it's not worth it, I will not push it further.

**Comment by fmassa** [comment]
I think one point that is important to keep in mind is consistency: you can take one model (for classification), feed the same input after the same transforms / normalization (or almost, see InceptionV3), and it will work, given you the predictions as expected. Adding models for CIFAR kind of breaks this consistency, because now inputs should have a different size / normalization.

Replicating original results from papers is very important, and ResNet is definitely a very important paper so being able to replicate it is definitely important. But I think replicating ResNets on CIFAR could be made available on a separate repo, maybe together with other implementations of models tailored specifically for CIFAR.

Let me know your thoughts

**Comment by nshaud** [comment]
@fmassa I was assuming that consistency was mostly important for pretrained models but maybe I'm wrong. You can already define weird ResNets using the existing class that are not available as pretrained models. I don't see the harm (from a user standpoint) of adding the missing variant from the original paper.

Anyway, to keep things separate, we could:
1. Add the missing variant to the torch.models.ResNet class without changing the available model list
2. Add the pretrained models and `models.narrow_resnets` architectures either in a `torch.models.cifar` or in a separate repo ("à la" sklearn-contrib) as you were suggesting.
